### PR TITLE
Indicator constraints in MOI

### PR DESCRIPTION
**Pull request by matbesancon**
Supports indicator constraints. The removal of these is not yet in the exposed C API (even though it is in CPLEX documentation), so it is simple additions for now. A new dict keeps track of the added indicator constraints

### CODE REVIEW

**Comment by matbesancon** [comment]
The tests are currently erroring. Checking for example `MOIT.indicator1_test`, it seems the indicator constraints are not registered. There is unfortunately no way of seeing this with CPLEX.

**Comment by odow** [comment]
> Indicator constraints are not registered. There is unfortunately no way of seeing this with CPLEX.

What do you mean by this?

**Comment by matbesancon** [comment]
The addition of the indicator constraints do not return any error, but the problem result is as is they were not added

**Comment by odow** [comment]
> but the problem result is as is they were not added

You probably need to set `model.has_int = true` in `add_indicator_constraint` since CPLEX has different discrete and continuous solution routines: https://github.com/JuliaOpt/CPLEX.jl/blob/2fe973549f0bba4f7b011fa07598185e34736ea1/src/cpx_solve.jl#L3

**Comment by matbesancon** [comment]
ok the solution was found in tears and blood, there was an off-by-one error in the cpx call, see the fix in `src/cpx_constrs.jl`. Also, the CPLEX documentation uses an Int32 for the sense instead of Char.

All indicator tests are now passing

**Comment by matbesancon** [comment]
Regarding `model.has_int = true`, it is supposed to be set when the binary variable is added. It is up to the user in all solvers to check that the indicator variable provided has been set to bin or integer and bounded

File: src/MOI/indicator_constraint.jl [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    if info_tuple === nothing
+        return false
+    end
+    return typeof(info_tuple[1].set) == S
+end
+
+function MOI.get(model::Optimizer,
+                 ::MOI.ConstraintSet,
+                 c::MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, MOI.IndicatorSet{A, S}}
+) where {A, S}
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Sorry to be a pain, but can we have consistent formatting? Also, we prefer to drop type parameters if not needed in the function.
```Julia
function MOI.get(
    model::Optimizer,
    ::MOI.ConstraintSet,
    c::MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, <:MOI.IndicatorSet}
)
```

File: src/MOI/indicator_constraint.jl [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    MOI.throw_if_not_valid(model, c)
+    info = model.indicator_constraint_info[c.value][1]
+    return info.set
+end
+
+function MOI.get(model::Optimizer,
+                 ::MOI.ConstraintFunction,
+                 c::MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, MOI.IndicatorSet{A, S}}
+) where {A, S}
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Ditto formatting

File: src/MOI/indicator_constraint.jl [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    MOI.throw_if_not_valid(model, c)
+    func = model.indicator_constraint_info[c.value][2]
+    return func
+end
+
+function MOI.add_constraint(
+        model::Optimizer, func::MOI.VectorAffineFunction{Float64},
+        s::MOI.IndicatorSet{A, S}) where {A, S <: Union{MOI.GreaterThan{Float64}, MOI.LessThan{Float64}, MOI.EqualTo{Float64}}}
+
+
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]:
```Julia
function MOI.add_constraint(
    model::Optimizer, 
    func::MOI.VectorAffineFunction{Float64},
    s::MOI.IndicatorSet{A, S}
) where {
    A, 
    S <: Union{MOI.GreaterThan{Float64}, MOI.LessThan{Float64}, MOI.EqualTo{Float64}}
}
```

And remove the redundant empty lines.

File: src/MOI/indicator_constraint.jl [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    first_index_terms  = [v.scalar_term for v in func.terms if v.output_index == 1]
+    scalar_index_terms = [v.scalar_term for v in func.terms if v.output_index != 1]
+    length(first_index_terms) == 1 || error("There should be exactly one term in output_index 1, found $(length(first_index_terms))")
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Prefer 
```Julia
if length(first_index_terms) != 1
    error()
end
```

File: src/MOI/indicator_constraint.jl [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            nindices += 1
+        end
+    end
+    return nindices
+end
+
+function MOI.get(model::Optimizer,
+                 ::MOI.ConstraintName,
+                 c::MOI.ConstraintIndex{<:MOI.VectorAffineFunction, S}
+) where {S <: MOI.IndicatorSet}
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Formatting. Also Prefer `MOI.ConstraintIndex{<:MOI.VectorAffineFunction, <:MOI.IndicatorSet}`.

File: src/MOI/indicator_constraint.jl [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    MOI.throw_if_not_valid(model, c)
+    info = model.indicator_constraint_info[c.value][1]
+    return info.name
+end
+
+function MOI.set(model::Optimizer,
+                 ::MOI.ConstraintName,
+                 c::MOI.ConstraintIndex{<:MOI.VectorAffineFunction, S},
+                 name::String,
+) where {S <: MOI.IndicatorSet}
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Formatting.

File: src/MOI/indicator_constraint.jl [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    MOI.throw_if_not_valid(model, c)
+    info = model.indicator_constraint_info[c.value][1]
+    info.name = name
+    return nothing
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: JuMP style guide prefers `return` over `return nothing`.

File: test/MathOptInterface/MOI_wrapper.jl [hunk 8]
```diff
@@ -76,10 +80,10 @@ end
 end
 
 @testset "Integer Linear tests" begin
-    MOIT.intlineartest(BRIDGED_OPTIMIZER, CONFIG, [
-        # TODO(odow): Indicator sets not supported.
-        "indicator1", "indicator2", "indicator3", "indicator4"
-    ])
+    # interval somehow needed for indicator tests
+    interval_optimizer = MOIB.LazyBridgeOptimizer(OPTIMIZER)
+    MOIB.add_bridge(interval_optimizer, MOIB.Constraint.SplitIntervalBridge{Float64})
```
Comment (odow) [inline test/MathOptInterface/MOI_wrapper.jl]: Remove the above three lines. We just need `BRIDGED_OPTIMIZER` below.

File: test/MathOptInterface/MOI_wrapper.jl [hunk 9]
```diff
@@ -76,10 +80,10 @@ end
...
+    MOIT.intlineartest(BRIDGED_OPTIMIZER, CONFIG, String[])
```
Comment (odow) [inline test/MathOptInterface/MOI_wrapper.jl]: Remove unneeded `String[]` as final argument.

File: src/MOI/indicator_constraint.jl [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+function MOI.get(
+    model::Optimizer, ::MOI.ListOfConstraintIndices{MOI.VectorAffineFunction{Float64}, S}
+) where {S <: MOI.IndicatorSet}
+    indices = MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, S}[]
+    for (key, info_func) in model.indicator_constraint_info
+        info = info_func[1]
+        if typeof(info.set) === S
+            push!(indices, MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, S}(key))
+        end
+    end
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Alternatively:
```Julia
indices = MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, S}[
    MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, S}(key)
    for (key, info_func) in model.indicator_constraint_info
    if typeof(info_func[1].set) == s
]

File: src/MOI/indicator_constraint.jl [hunk 11]
```diff
@@ -0,0 +1,117 @@
+
+function MOI.supports_constraint(
+    ::Optimizer, ::Type{MOI.VectorAffineFunction{Float64}}, ::Type{<:MOI.IndicatorSet{A, S}}
+) where {A, S <: Union{MOI.LessThan{Float64}, MOI.GreaterThan{Float64}, MOI.EqualTo{Float64}}}
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Is `EqualTo` tested? 

Is there are a restriction on `A`?

File: src/MOI/indicator_constraint.jl [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    return true
+end
+
+function MOI.is_valid(
+    model::Optimizer,
+    c::MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, S}
+) where {A, S <: MOI.IndicatorSet}
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Is this correct? What is `A` here?

File: src/MOI/indicator_constraint.jl (hunk 1 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: what formatting do you refer to, the parenthesis on a new line?

File: src/MOI/indicator_constraint.jl (hunk 5 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: We need the exact type of `S` to construct `indices = MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, S}[]` below

Comment (odow) [reply src/MOI/indicator_constraint.jl]: Not in this function?

File: src/MOI/indicator_constraint.jl (hunk 1 above)
Comment (odow) [reply src/MOI/indicator_constraint.jl]: Dropping arguments to a new line and indenting 4 spaces, with parenthesis on new line.

File: test/MathOptInterface/MOI_wrapper.jl (hunk 8 above)
Comment (matbesancon) [reply test/MathOptInterface/MOI_wrapper.jl]: `BRIDGED_OPTIMIZER` is using all bridges, including the bridge from indicator to SOS, the test could thus pass without testing the real implementation if an error is made (like `MOI.supports_constraint` returning the wrong type

Comment (odow) [reply test/MathOptInterface/MOI_wrapper.jl]: Okay, but then we need to test the indicator tests with `interval_optimizer`. Currently, it is un-used.

File: src/MOI/indicator_constraint.jl (hunk 11 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: See the docstring for `MOI.IndicatorSet{A, S}`
A is a value from the enum `MOI.ActivationCondition`.:
```
julia> typeof(MOI.ACTIVATE_ON_ONE)
Enum MathOptInterface.ActivationCondition:
ACTIVATE_ON_ZERO = 0
ACTIVATE_ON_ONE = 1
```

Unfortunately, one cannot type-check value parameters of a type

Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: > Is EqualTo tested?

No that's a good point, it is missing from MOI, I'll add a PR with an equality test

File: src/MOI/indicator_constraint.jl (hunk 12 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: same as above, `A` is the MOI.ActivationCondition

File: src/MOI/indicator_constraint.jl (hunk 5 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: correct, I'm updating that too

File: src/MOI/indicator_constraint.jl (hunk 12 above)
Comment (odow) [reply src/MOI/indicator_constraint.jl]: But `A` isn't used in the function, or in the header. You probably just mean `) where {S <: IndicatorSet}`?

Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: that's right, I hadn't caught this one thanks

File: test/MathOptInterface/MOI_wrapper.jl (hunk 8 above)
Comment (matbesancon) [reply test/MathOptInterface/MOI_wrapper.jl]: totally correct, just fixed that

File: src/MOI/indicator_constraint.jl (hunk 7 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: done

File: src/MOI/indicator_constraint.jl (hunk 1 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: addressed in last commit

**Comment by matbesancon** [comment]
bump there

File: src/MOI/MOI_wrapper.jl [hunk 13]
```diff
@@ -305,6 +309,8 @@ function MOI.supports_constraint(
     # Note: CPLEX does not support quadratic equality constraints.
     return true
 end
+import MathOptInterface
+const MOI = MathOptInterface
```
Comment (odow) [inline src/MOI/MOI_wrapper.jl]: Remove.

File: src/MOI/indicator_constraint.jl [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    # a^T x + b <= c <==> a^T x <= c - b
+    fcons = MOI.constant(func)
+    rhs = Cdouble(MOI.constant(s.set) - fcons[2])
+    add_indicator_constraint(model.inner, linear_idx, linear_coefficients, sense_code, rhs, var_inner, active_bool)
+    return MOI.ConstraintIndex{typeof(func), typeof(s)}(model.last_constraint_index)
+end
+
+_get_indicator_sense(::MOI.GreaterThan) = Cint('G')
+_get_indicator_sense(::MOI.LessThan) = Cint('L')
+_get_indicator_sense(::MOI.EqualTo) = Cint('E')
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Is `Cint` correct here?!? Not `Cchar`?

File: src/MOI/indicator_constraint.jl [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+function MOI.get(
+        model::Optimizer,
+        ::MOI.ListOfConstraintIndices{<:MOI.VectorAffineFunction, S},
+) where {
+    S <: MOI.IndicatorSet
+}
+
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Remove blank line

File: src/MOI/indicator_constraint.jl [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    ::Type{<:MOI.IndicatorSet{A, S}}
+) where {
+    A, S <: Union{MOI.LessThan{Float64}, MOI.GreaterThan{Float64}, MOI.EqualTo{Float64}}
+}
+    return true
+end
+
+function MOI.is_valid(
+        model::Optimizer,
+        c::MOI.ConstraintIndex{MOI.VectorAffineFunction{Float64}, S}
```
Comment (odow) [inline src/MOI/indicator_constraint.jl]: Here and throughout, indentation is off.

File: src/cpx_constrs.jl [hunk 17]
```diff
@@ -494,12 +494,12 @@ function add_indicator_constraint(model::Model, idx::Vector{Cint}, coeff::Vector
                       Cint,
                       Cint,
                       Cdouble,
-                      Cchar,
+                      Cint,
```
Comment (odow) [inline src/cpx_constrs.jl]: I understand that the [docs say this is a `Cint`](https://www.ibm.com/support/knowledgecenter/SSSA5P_12.10.0/ilog.odms.cplex.help/refcallablelibrary/gcapi/addindconstr.html), but this is surely a mistake. What would this be the only place in the API where the sense is a `Cint`. What happens if you make this a `Cchar`? Does it work? 

I can follow up with the developers.

File: src/MOI/indicator_constraint.jl (hunk 14 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: Yes, I don't know why the signature in the wrapper takes a Cchar:
https://www.ibm.com/support/knowledgecenter/SSSA5P_12.9.0/ilog.odms.cplex.help/refcallablelibrary/gcapi/addindconstr.html
this is specified as int, might have been a char in a past version

File: src/cpx_constrs.jl (hunk 17 above)
Comment (matbesancon) [reply src/cpx_constrs.jl]: I'll re-test with Cchar to see if it works

Comment (odow) [reply src/cpx_constrs.jl]: I've mailed the CPLEX developers for clarification.

Comment (matbesancon) [reply src/cpx_constrs.jl]: seems all fine, I reverted to Cchar

File: src/MOI/indicator_constraint.jl (hunk 16 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: changed in last commit

File: src/MOI/indicator_constraint.jl (hunk 14 above)
Comment (matbesancon) [reply src/MOI/indicator_constraint.jl]: Converted to Cchar in last commit

File: src/MOI/MOI_wrapper.jl (hunk 13 above)
Comment (matbesancon) [reply src/MOI/MOI_wrapper.jl]: done in last commit

File: src/cpx_constrs.jl (hunk 17 above)
Comment (odow) [reply src/cpx_constrs.jl]: Here's the response from CPLEX:

image

Please change to `Cint` and leave a comment.

Comment (matbesancon) [reply src/cpx_constrs.jl]: done thanks for reaching out to them

**Comment by matbesancon** [comment]
this is all good on my side, I think the MOI incompatibility should be fixed in another PR

**Comment by matbesancon** [comment]
the tests are passing with MOI v0.9.11

**Comment by matbesancon** [comment]
bump @odow

**Comment by odow** [comment]
Screenshot of the tests passing?

**Comment by matbesancon** [comment]
image

image

**Pull request merged**